src/NuGet.Core/NuGet.Versioning/VersionComparer.cs

```diff
@@ -68,16 +68,19 @@ public bool Equals(SemanticVersion x, SemanticVersion y)
         /// <summary>
         /// Compares the given versions using the VersionComparison mode.
         /// </summary>
-        public static int Compare(SemanticVersion version1, SemanticVersion version2, VersionComparison versionComparison)
+        public static int Compare(SemanticVersion? version1, SemanticVersion? version2, VersionComparison versionComparison)
         {
             IVersionComparer comparer = new VersionComparer(versionComparison);
+#pragma warning disable CS8604 // Possible null reference argument.
+            // The BCL is missing nullable annotations on IComparable<T> before net5.0
             return comparer.Compare(version1, version2);
+#pragma warning restore CS8604 // Possible null reference argument.
         }
 
         /// <summary>
         /// Gives a hash code based on the normalized version string.
         /// </summary>
-        public int GetHashCode(SemanticVersion version)
+        public int GetHashCode(SemanticVersion? version)
```

Comment:
Do we know why GetHashCode has no annotations? 
I'd lean towards not annotating to match the behavior if the reasoning is good.

Reply:
[The docs for `IEqualityComparer<T>.GetHashCode` say](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.iequalitycomparer-1.gethashcode?view=net-7.0)

> Exceptions
> [ArgumentNullException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentnullexception?view=net-7.0)
> The type of obj is a reference type and obj is null.

My guess is that it's like asking what colour a non-existent ball is. It doesn't fundamentally make sense to ask what the hash code for a null value is. Our implementation returns `0` for this scenario, but `0` is a perfectly valid hash code for a non-null value to have, so now there's a potential hash collision between null and non-null values (although there's also potentially hash collisions for other non-null values, which is why it's critical to check equality and don't assume hash codes are unique).

I guess this is also the reason why `Dictionary<T1, T2>` doesn't allow null keys (even though there are times where that would be useful).

I'll remove the nullable annotation, but I won't change the implementation (to throw an exception, rather than return 0) for now.